annotate pod  policy.cilium.io/proxy-visibility causing cilium to panic with WireGuard

### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Before 1.14, we can't have l7 proxy with WireGuard. we prevent it from in here https://github.com/cilium/cilium/blob/v1.13/daemon/cmd/daemon_main.go#L1604

However, there is a loophole. if we directly annotate the pod for l7 visibility, that will directly panic cilium. The cilium pod will be in crashloop . I think we should prevent it from panic
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x21b1f90]

goroutine 939 [running]:
github.com/cilium/cilium/pkg/proxy.(*Proxy).CreateOrUpdateRedirect(0x0, {0x31f9e38?, 0x4000f42200}, {0x31fe020, 0x400006f440}, {0x4000efcca8, 0x12}, {0x320fc20, 0x4000f3ce00}, 0x4000191940)
        /go/src/github.com/cilium/cilium/pkg/proxy/proxy.go:491 +0x60
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).addVisibilityRedirects(0x4000f3ce00, 0x0, 0x2c414d0?, 0x6?)
        /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:390 +0x35c
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).addNewRedirects(0x4000f3ce00, 0x4000edd680?)
        /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:476 +0x308
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).runPreCompilationSteps(0x4000f3ce00, 0x400137c400, 0x0)
        /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:853 +0x3cc
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerateBPF(0x4000f3ce00, 0x400137c400)
        /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:594 +0x160
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerate(0x4000f3ce00, 0x400137c400)
        /go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:457 +0x600
github.com/cilium/cilium/pkg/endpoint.(*EndpointRegenerationEvent).Handle(0x400001e320, 0x40009c0f28?)
        /go/src/github.com/cilium/cilium/pkg/endpoint/events.go:53 +0x2a4
github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).run.func1()
        /go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:245 +0x108
sync.(*Once).doSlow(0x0?, 0x7e098?)
        /usr/local/go/src/sync/once.go:74 +0x104
sync.(*Once).Do(...)
        /usr/local/go/src/sync/once.go:65
github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).run(0x40009dcd70?)
        /go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:233 +0x48
created by github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).Run
        /go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:229 +0x7c
```

### Cilium Version

v1.13.6 v.1.14.1

### Kernel Version

Linux minikube 5.15.0-78-generic #85-Ubuntu SMP Fri Jul 7 15:29:30 UTC 2023 aarch64 aarch64 aarch64 GNU/Linux

### Kubernetes Version

I tested with 1.24.15

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

you can just use the following helm values
```
cluster:
  name: minikube
encryption:
  enabled: true
  nodeEncryption: false
  type: wireguard
endpointStatus:
  enabled: true
  status: policy
l7Proxy: false
operator:
  replicas: 1
serviceAccounts:
  cilium:
    name: cilium
  operator:
    name: cilium-operator
tunnel: vxlan
```
then run 2 nginx pods
```
k run nginx --image=nginx
kubectl annotate pod nginx  policy.cilium.io/proxy-visibility="<Egress/53/UDP/DNS>,<Egress/80/TCP/HTTP>"
k run nginx1 --image=nginx
```
the nginx1 will cause the cilium to crash


### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/Documentation/observability/visibility.rst
+++ b/Documentation/observability/visibility.rst
@@ -10,6 +10,10 @@
 Layer 7 Protocol Visibility
 ***************************
 
+.. note::
+
+    This feature requires enabling L7 Proxy support. Without it, the visibility annotation is ignored.
+
 While :ref:`monitor` provides introspection into datapath state, by default it
 will only provide visibility into L3/L4 packet events. If :ref:`l7_policy` are
 configured, one can get visibility into L7 protocols, but this requires the full

--- a/daemon/cmd/endpoint.go
+++ b/daemon/cmd/endpoint.go
@@ -475,7 +475,7 @@ func (d *Daemon) createEndpoint(ctx context.Context, owner regeneration.Owner, e
 		return d.errorDuringCreation(ep, fmt.Errorf("unable to insert endpoint into manager: %s", err))
 	}
 
-	// We need to update the the visibility policy after adding the endpoint in
+	// We need to update the visibility policy after adding the endpoint in
 	// the endpoint manager because the endpoint manager create the endpoint
 	// queue of the endpoint. If we execute this function before the endpoint
 	// manager creates the endpoint queue the operation will fail.
@@ -488,6 +488,7 @@ func (d *Daemon) createEndpoint(ctx context.Context, owner regeneration.Owner, e
 			value, _ := annotation.Get(p, annotation.ProxyVisibility, annotation.ProxyVisibilityAlias)
 			return value, nil
 		})
+
 		ep.UpdateBandwidthPolicy(func(ns, podName string) (bandwidthEgress string, err error) {
 			_, p, err := d.endpointMetadataFetcher.Fetch(ns, podName)
 			if err != nil {

--- a/install/kubernetes/cilium/templates/validate.yaml
+++ b/install/kubernetes/cilium/templates/validate.yaml
@@ -48,10 +48,10 @@
   {{- end }}
 {{- end }}
 
-{{- if or .Values.ingressController.enabled .Values.gatewayAPI.enabled }}
+{{- if or .Values.ingressController.enabled .Values.gatewayAPI.enabled (eq .Values.loadBalancer.l7.backend "envoy") }}
   {{- if hasKey .Values "l7Proxy" }}
     {{- if not .Values.l7Proxy }}
-      {{ fail "Ingress or Gateway API controller requires .Values.l7Proxy to be set to 'true'" }}
+      {{ fail "Ingress or Gateway API controller or Envoy L7 Load Balancer  requires .Values.l7Proxy to be set to 'true'" }}
     {{- end }}
   {{- end }}
 {{- end }}

--- a/pkg/endpoint/bpf.go
+++ b/pkg/endpoint/bpf.go
@@ -202,7 +202,7 @@ func (e *Endpoint) GetLabelsLocked(id identity.NumericIdentity) labels.LabelArra
 // problem that occurred while adding an l7 redirect for the specified policy.
 // Must be called with endpoint.mutex Lock()ed.
 func (e *Endpoint) addNewRedirectsFromDesiredPolicy(ingress bool, desiredRedirects map[string]bool, proxyWaitGroup *completion.WaitGroup) (error, revert.FinalizeFunc, revert.RevertFunc) {
-	if option.Config.DryMode || e.isProxyDisabled() {
+	if option.Config.DryMode || e.IsProxyDisabled() {
 		return nil, nil, nil
 	}
 
@@ -310,7 +310,7 @@ func (e *Endpoint) addVisibilityRedirects(ingress bool, desiredRedirects map[str
 		}
 	)
 
-	if e.visibilityPolicy == nil {
+	if e.visibilityPolicy == nil || e.IsProxyDisabled() {
 		return nil, finalizeList.Finalize, revertStack.Revert
 	}
 

--- a/pkg/endpoint/events.go
+++ b/pkg/endpoint/events.go
@@ -226,6 +226,15 @@ func (ev *EndpointPolicyVisibilityEvent) Handle(res chan interface{}) {
 		return
 	}
 	if proxyVisibility != "" {
+		if e.IsProxyDisabled() {
+			e.getLogger().
+				WithField(logfields.EndpointID, e.GetID()).
+				Warn("ignoring L7 proxy visibility policy as L7 proxy is disabled")
+			res <- &EndpointRegenerationResult{
+				err: nil,
+			}
+			return
+		}
 		e.getLogger().Debug("creating visibility policy")
 		nvp, err = policy.NewVisibilityPolicy(proxyVisibility)
 		if err != nil {

--- a/pkg/endpoint/policy.go
+++ b/pkg/endpoint/policy.go
@@ -121,7 +121,7 @@ func (e *Endpoint) updateNetworkPolicy(proxyWaitGroup *completion.WaitGroup) (re
 		return nil, nil
 	}
 
-	if e.isProxyDisabled() {
+	if e.IsProxyDisabled() {
 		return nil, nil
 	}
 

--- a/pkg/endpoint/policy_test.go
+++ b/pkg/endpoint/policy_test.go
@@ -30,7 +30,7 @@ import (
 
 func (s *EndpointSuite) TestUpdateVisibilityPolicy(c *check.C) {
 	do := &DummyOwner{repo: policy.NewPolicyRepository(nil, nil, nil, nil)}
-	ep := NewEndpointWithState(do, do, testipcache.NewMockIPCache(), nil, testidentity.NewMockIdentityAllocator(nil), 12345, StateReady)
+	ep := NewEndpointWithState(do, do, testipcache.NewMockIPCache(), &FakeEndpointProxy{}, testidentity.NewMockIdentityAllocator(nil), 12345, StateReady)
 	ep.UpdateVisibilityPolicy(func(_, _ string) (string, error) {
 		return "", nil
 	})

--- a/pkg/endpoint/proxy.go
+++ b/pkg/endpoint/proxy.go
@@ -29,13 +29,13 @@ func (e *Endpoint) SetProxy(p EndpointProxy) {
 }
 
 func (e *Endpoint) removeNetworkPolicy() {
-	if e.isProxyDisabled() {
+	if e.IsProxyDisabled() {
 		return
 	}
 	e.proxy.RemoveNetworkPolicy(e)
 }
 
-func (e *Endpoint) isProxyDisabled() bool {
+func (e *Endpoint) IsProxyDisabled() bool {
 	return e.proxy == nil || reflect.ValueOf(e.proxy).IsNil()
 }
 

